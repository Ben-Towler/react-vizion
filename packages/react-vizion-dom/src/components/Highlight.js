import React from 'react';
import { createClassNames } from '../core/utils';
import Highlighter from './Highlighter';

const cx = createClassNames('highlight');

const Highlight = props => (
  <Highlighter {...props} highlightProperty="_highlight" cx={cx} />
);

export default Highlight;
