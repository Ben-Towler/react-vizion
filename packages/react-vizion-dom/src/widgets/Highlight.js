import { connectHighlight } from '@clinia/react-vizion-core';
import Highlight from '../components/Highlight';

/**
 * Renders any property from a hit into its highlighted form when relevant.
 *
 * @name Highlight
 * @kind widget
 * @propType {string} property - location of the highlighted property in the hit (the corresponding element can be either a string or an array of strings)
 * @propType {object} hit - hit object containing the highlighted property
 * @propType {string} [tagName='em'] - tag to be used for highlighted parts of the hit
 * @propType {string} [nonHighlightedTagName='span'] - tag to be used for the parts of the hit that are not highlighted
 * @propType {node} [separator=',<space>'] - symbol used to separate the elements of the array in case the property points to an array of strings.
 * @themeKey cvi-highlight - root of the component
 * @themeKey cvi-highlight-highlighted - part of the text which is highlighted
 * @themeKey cvi-highlight-nonHighlighted - part of the text that is not highlighted
 * @example
 * import React from 'react';
 * import clinia from 'clinia/lite';
 * import { Vizion, SearchBox, Hits, Highlight } from '@clinia/react-vizion-dom';
 *
 * const Hit = ({ hit }) => (
 *   <div>
 *     <Highlight property="name" hit={hit} />
 *   </div>
 * );
 *
 * const searchClient = clinia(
 *   'latency',
 *   '6be0576ff61c053d5f9a3225e2a90f76'
 * );
 *
 * const App = () => (
 *   <Vizion
 *     searchClient={searchClient}
 *     indexName="instant_search"
 *   >
 *     <SearchBox defaultRefinement="Pho" />
 *     <Hits hitComponent={Hit} />
 *   </Vizion>
 * );
 */

export default connectHighlight(Highlight);
