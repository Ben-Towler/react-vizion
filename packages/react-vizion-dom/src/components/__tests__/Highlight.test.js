import React from 'react';
import renderer from 'react-test-renderer';
import Highlight from '../Highlight';

describe('Highlight', () => {
  it('parses an highlighted property of hit object', () => {
    const hitFromAPI = {
      id: 0,
      deep: { property: { value: 'awesome highlighted hit!' } },
      _highlight: {
        deep: {
          property: {
            value: {
              value:
                'awesome <cvi-highlight>hi</cvi-highlight>ghlighted <cvi-highlight>hi</cvi-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = () => [
      { value: 'awesome ', isHighlighted: false },
      { value: 'hi', isHighlighted: true },
      { value: 'ghlighted ', isHighlighted: false },
      { value: 'hi', isHighlighted: true },
      { value: 't!', isHighlighted: false },
    ];

    const tree = renderer.create(
      <Highlight
        cx={(...x) => x.join(' ')}
        property="deep.property.value"
        hit={hitFromAPI}
        highlight={highlight}
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders a hit with a custom tag correctly', () => {
    const hitFromAPI = {
      objectID: 0,
      deep: { property: { value: 'awesome highlighted hit!' } },
      _highlight: {
        deep: {
          property: {
            value: {
              value:
                'awesome <cvi-highlight>hi</cvi-highlight>ghlighted <cvi-highlight>hi</cvi-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = () => [
      { value: 'awesome ', isHighlighted: false },
      { value: 'hi', isHighlighted: true },
      { value: 'ghlighted ', isHighlighted: false },
      { value: 'hi', isHighlighted: true },
      { value: 't!', isHighlighted: false },
    ];

    const tree = renderer.create(
      <Highlight
        cx={(...x) => x.join(' ')}
        property="deep.property.value"
        hit={hitFromAPI}
        highlight={highlight}
        tagName="marquee"
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders a hit with a custom className', () => {
    const hitFromAPI = {
      id: 0,
      deep: { property: { value: 'awesome highlighted hit!' } },
      _highlight: {
        deep: {
          property: {
            value: {
              value:
                'awesome <cvi-highlight>hi</cvi-highlight>ghlighted <cvi-highlight>hi</cvi-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = () => [
      { value: 'awesome ', isHighlighted: false },
      { value: 'hi', isHighlighted: true },
      { value: 'ghlighted ', isHighlighted: false },
      { value: 'hi', isHighlighted: true },
      { value: 't!', isHighlighted: false },
    ];

    const tree = renderer.create(
      <Highlight
        cx={(...x) => x.join(' ')}
        className="MyCustomHighlight"
        property="deep.property.value"
        hit={hitFromAPI}
        highlight={highlight}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
