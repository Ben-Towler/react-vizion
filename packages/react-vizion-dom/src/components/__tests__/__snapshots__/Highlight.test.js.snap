// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlight parses an highlighted property of hit object 1`] = `
<span
  className="cvi-highlight"
>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    awesome 
  </span>
  <em
    className="cvi-highlight-highlighted"
  >
    hi
  </em>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    ghlighted 
  </span>
  <em
    className="cvi-highlight-highlighted"
  >
    hi
  </em>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    t!
  </span>
</span>
`;

exports[`Highlight renders a hit with a custom className 1`] = `
<span
  className="cvi-highlight MyCustomHighlight"
>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    awesome 
  </span>
  <em
    className="cvi-highlight-highlighted"
  >
    hi
  </em>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    ghlighted 
  </span>
  <em
    className="cvi-highlight-highlighted"
  >
    hi
  </em>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    t!
  </span>
</span>
`;

exports[`Highlight renders a hit with a custom tag correctly 1`] = `
<span
  className="cvi-highlight"
>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    awesome 
  </span>
  <marquee
    className="cvi-highlight-highlighted"
  >
    hi
  </marquee>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    ghlighted 
  </span>
  <marquee
    className="cvi-highlight-highlighted"
  >
    hi
  </marquee>
  <span
    className="cvi-highlight-nonHighlighted"
  >
    t!
  </span>
</span>
`;
